llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -1678,6 +1694,12 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
       }
     }
 
+    // GFX11 must wait for any pending GDS instruction to complete before
+    // any "Always GDS" instruction.
+    if (AMDGPU::isGFX11(*ST) && TII->isAlwaysGDS(MI.getOpcode()) &&
```

Comment:
Don't restrict this to GFX11. The rule definitely applies to GFX9 and GFX10 too, so I'm guessing it applies to any architecture that supports GWS and this code can be unconditional.